clients/src/main/java/org/apache/kafka/clients/consumer/internals/NetworkClientDelegate.java

```diff
@@ -0,0 +1,287 @@
+import org.apache.kafka.common.utils.LogContext;
+import org.apache.kafka.common.utils.Time;
+import org.apache.kafka.common.utils.Timer;
+import org.slf4j.Logger;
+
+import java.io.IOException;
+import java.util.ArrayDeque;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Queue;
+import java.util.Set;
+
+/**
+ * A wrapper around the {@link org.apache.kafka.clients.NetworkClient} to handle poll and send operations.
+ */
+public class NetworkClientDelegate implements AutoCloseable {
+    private final KafkaClient client;
+    private final Time time;
+    private final Logger log;
+    private final int requestTimeoutMs;
+    private boolean wakeup = false;
+    private final Queue<UnsentRequest> unsentRequests;
+    private final Set<Node> activeNodes;
+
+    public NetworkClientDelegate(
+            final Time time,
+            final ConsumerConfig config,
+            final LogContext logContext,
+            final KafkaClient client) {
+        this.time = time;
+        this.client = client;
+        this.log = logContext.logger(getClass());
+        this.unsentRequests = new ArrayDeque<>();
+        this.requestTimeoutMs = config.getInt(ConsumerConfig.REQUEST_TIMEOUT_MS_CONFIG);
+        this.activeNodes = new HashSet<>();
+    }
+
+    /**
+     * Returns the responses of the sent requests. This methods will try to send the unsent requests, poll for responses,
+     * and check the disconnected nodes.
+     * @param timeoutMs
+     * @return
+     */
+    public List<ClientResponse> poll(final long timeoutMs) {
+        final long currentTimeMs = time.milliseconds();
+        trySend(currentTimeMs);
+        List<ClientResponse> res = this.client.poll(timeoutMs, currentTimeMs);
+        checkDisconnects();
+        maybeTriggerWakeup();
+        return res;
+    }
+
+    /**
+     * Iterates through the unsentRequests queue and tries to send the unsent requests. If the request doesn't have an
+     * assigned node, it will find the leastLoadedOne.
+     * Here it also stores all the nodes in the {@code activeNodes}, which will then be used to check for the disconnection.
+     */
+    void trySend(final long currentTimeMs) {
+        Queue<UnsentRequest> unsentAndUnreadyRequests = new LinkedList<>();
+        while (!unsentRequests.isEmpty()) {
+            UnsentRequest unsent = unsentRequests.poll();
+            unsent.timer.update(currentTimeMs);
+            if (unsent.timer.isExpired()) {
+                unsent.callback.ifPresent(c -> c.onFailure(new TimeoutException(
+                        "Failed to send request after " + unsent.timer.timeoutMs() + " " + "ms.")));
+                continue;
+            }
+
+            if (!doSend(unsent, currentTimeMs, unsentAndUnreadyRequests)) {
+                log.debug("No broker available to send the request: {}", unsent);
```

Comment:
Can't we retry internally until the request expires? What is the advantage to propagating this back at this point?